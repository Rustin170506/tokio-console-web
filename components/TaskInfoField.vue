<template>
    <div class="flex">
        <span class="font-bold text-black dark:text-white">{{ name }}</span>
        <span class="mr-1">: </span>
        <slot
            ><span :class="valueClass">{{ value }}</span></slot
        >
    </div>
</template>
<script>
export default {
    props: {
        name: {
            type: String,
            required: true,
        },
        value: {
            type: [Number, String, BigInt, DurationWithStyle],
            default: null,
        },
    },
    computed: {
        valueClass() {
            return this.value instanceof DurationWithStyle
                ? this.value.class
                : "";
        },
    },
};
</script>
