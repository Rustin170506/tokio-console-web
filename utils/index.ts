import type { Ref } from "vue";
import { getCurrentScope, onScopeDispose, ref, watch } from "vue";

export function useObjectStorage<T>(
    key: string,
    initial: T,
    listenToStorage = true,
): Ref<T> {
    const raw = localStorage.getItem(key);
    const data = ref(raw ? JSON.parse(raw) : initial);

    for (const key in initial) {
        if (data.value[key] === undefined) data.value[key] = initial[key];
    }

    let updating = false;
    let wrote = "";

    watch(
        data,
        (value) => {
            if (updating) return;
            wrote = JSON.stringify(value);
            localStorage.setItem(key, wrote);
        },
        { deep: true, flush: "post" },
    );

    if (listenToStorage) {
        useEventListener(window, "storage", (e: StorageEvent) => {
            if (e.key === key && e.newValue && e.newValue !== wrote) {
                updating = true;
                data.value = JSON.parse(e.newValue);
                updating = false;
            }
        });
    }

    return data;
}

export function useEventListener(
    target: EventTarget,
    type: string,
    listener: any,
    options?: boolean | AddEventListenerOptions,
) {
    target.addEventListener(type, listener, options);
    getCurrentScope() &&
        onScopeDispose(() =>
            target.removeEventListener(type, listener, options),
        );
}

const topVarName = "--console-safe-area-top";
const rightVarName = "--console-safe-area-right";
const bottomVarName = "--console-safe-area-bottom";
const leftVarName = "--console-safe-area-left";

/**
 * Reactive `env(safe-area-inset-*)`
 *
 * @see https://vueuse.org/useScreenSafeArea
 */
export function useScreenSafeArea() {
    const top = ref(0);
    const right = ref(0);
    const bottom = ref(0);
    const left = ref(0);

    document.documentElement.style.setProperty(
        topVarName,
        "env(safe-area-inset-top, 0px)",
    );
    document.documentElement.style.setProperty(
        rightVarName,
        "env(safe-area-inset-right, 0px)",
    );
    document.documentElement.style.setProperty(
        bottomVarName,
        "env(safe-area-inset-bottom, 0px)",
    );
    document.documentElement.style.setProperty(
        leftVarName,
        "env(safe-area-inset-left, 0px)",
    );

    update();
    useEventListener(window, "resize", update);

    function getValue(position: string) {
        return (
            Number.parseFloat(
                getComputedStyle(document.documentElement).getPropertyValue(
                    position,
                ),
            ) || 0
        );
    }

    function update() {
        top.value = getValue(topVarName);
        right.value = getValue(rightVarName);
        bottom.value = getValue(bottomVarName);
        left.value = getValue(leftVarName);
    }

    return {
        top,
        right,
        bottom,
        left,
        update,
    };
}
