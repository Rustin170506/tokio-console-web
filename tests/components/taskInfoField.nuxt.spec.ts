import { expect, it } from "vitest";
import { renderSuspended } from "nuxt-vitest/utils";
import TaskInfoField from "~/components/TaskInfoField.vue";

it("TaskInfoField without slot", async () => {
    const component = await renderSuspended(TaskInfoField, {
        props: {
            name: "Test Name",
            value: "Test Value",
        },
    });
    expect(component.html()).toMatchSnapshot();
});

it("TaskInfoField with slot", async () => {
    const component = await renderSuspended(TaskInfoField, {
        props: {
            name: "Test Name",
        },
        slots: {
            default: () => "Test Slot Content",
        },
    });
    expect(component.html()).toMatchSnapshot();
});
