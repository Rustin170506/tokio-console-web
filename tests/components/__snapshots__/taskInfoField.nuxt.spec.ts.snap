// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InfoField with slot 1`] = `
"<div id=\\"test-wrapper\\">
  <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">Test Name</span><span class=\\"mr-1\\">: </span>Test Slot Content</div>
</div>"
`;

exports[`InfoField without slot 1`] = `
"<div id=\\"test-wrapper\\">
  <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">Test Name</span><span class=\\"mr-1\\">: </span><span class=\\"\\">Test Value</span></div>
</div>"
`;

exports[`TaskInfoField with slot 1`] = `
"<div id=\\"test-wrapper\\">
  <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">Test Name</span><span class=\\"mr-1\\">: </span>Test Slot Content</div>
</div>"
`;

exports[`TaskInfoField without slot 1`] = `
"<div id=\\"test-wrapper\\">
  <div class=\\"flex\\"><span class=\\"font-bold text-black dark:text-white\\">Test Name</span><span class=\\"mr-1\\">: </span><span class=\\"\\">Test Value</span></div>
</div>"
`;
